plugins/ca/root-ca/src/org/apache/cloudstack/ca/provider/RootCAProvider.java

```diff
@@ -0,0 +1,572 @@
+            LOG.error("Failed to read provided CSR string as a PEM object", e);
+        }
+
+        if (pemObject == null) {
+            throw new CloudRuntimeException("Unable to read/process CSR: " + csr);
+        }
+
+        final PKCS10CertificationRequest request = new PKCS10CertificationRequest(pemObject.getContent());
+
+        final X509Certificate clientCertificate = CertUtils.generateV3Certificate(
+                caCertificate, caKeyPair.getPrivate(),
+                request.getPublicKey(),
+                request.getCertificationRequestInfo().getSubject().toString(),
+                CAManager.CertSignatureAlgorithm.value(),
+                validityDays,
+                domainNames,
+                ipAddresses);
+        return new Certificate(clientCertificate, null, Collections.singletonList(caCertificate));
+
+    }
+
+    ////////////////////////////////////////////////////////
+    /////////////// Root CA API Handlers ///////////////////
+    ////////////////////////////////////////////////////////
+
+    @Override
+    public boolean canProvisionCertificates() {
+        return true;
+    }
+
+    @Override
+    public List<X509Certificate> getCaCertificate() {
+        return Collections.singletonList(caCertificate);
+    }
+
+    @Override
+    public Certificate issueCertificate(final List<String> domainNames, final List<String> ipAddresses, final int validityDays) {
+        try {
+            return generateCertificate(domainNames, ipAddresses, validityDays);
+        } catch (final CertificateException | IOException | SignatureException | NoSuchAlgorithmException | NoSuchProviderException | InvalidKeyException | OperatorCreationException e) {
+            LOG.error("Failed to create client certificate, due to: ", e);
+            throw new CloudRuntimeException("Failed to generate certificate due to:" + e.getMessage());
+        }
+    }
+
+    @Override
+    public Certificate issueCertificate(final String csr, final List<String> domainNames, final List<String> ipAddresses, final int validityDays) {
+        try {
+            return generateCertificateUsingCsr(csr, domainNames, ipAddresses, validityDays);
+        } catch (final CertificateException | IOException | SignatureException | NoSuchAlgorithmException | NoSuchProviderException | InvalidKeyException | OperatorCreationException e) {
+            LOG.error("Failed to generate certificate from CSR: ", e);
+            throw new CloudRuntimeException("Failed to generate certificate using CSR due to:" + e.getMessage());
+        }
+    }
+
+    @Override
+    public boolean revokeCertificate(final BigInteger certSerial, final String certCn) {
+        return true;
+    }
+
+    ////////////////////////////////////////////////////////////
+    /////////////// Root CA Trust Management ///////////////////
+    ////////////////////////////////////////////////////////////
+
+    public static final class RootCACustomTrustManager implements X509TrustManager {
```

Comment:
It needs to be restricted to the `RootCAProvider` class, however, I made this public to write unit tests easily. I'll see if we can refactor it out completely.